crates/sui-indexer-alt/src/handlers/sum_packages.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use diesel::{upsert::excluded, ExpressionMethods};
+use diesel_async::RunQueryDsl;
+use futures::future::try_join_all;
+use sui_types::full_checkpoint_content::CheckpointData;
+
+use crate::{
+    db,
+    models::packages::StoredPackage,
+    pipeline::{sequential::Handler, Processor},
+    schema::sum_packages,
+};
+
+const CHUNK_ROWS: usize = i16::MAX as usize / 5;
+
+pub struct SumPackages;
+
+impl Processor for SumPackages {
+    const NAME: &'static str = "sum_packages";
+
+    type Value = StoredPackage;
+
+    fn process(checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
+        let CheckpointData {
+            checkpoint_summary,
+            transactions,
+            ..
+        } = checkpoint.as_ref();
+
+        let cp_sequence_number = checkpoint_summary.sequence_number as i64;
+        let mut values = vec![];
+        for tx in transactions {
+            for obj in &tx.output_objects {
+                let Some(package) = obj.data.try_as_package() else {
+                    continue;
+                };
+
+                values.push(StoredPackage {
+                    package_id: obj.id().to_vec(),
+                    original_id: package.original_package_id().to_vec(),
+                    package_version: obj.version().value() as i64,
+                    move_package: bcs::to_bytes(package)
+                        .map_err(|e| anyhow!("Error serializing package {}: {e}", obj.id()))?,
+                    cp_sequence_number,
+                });
+            }
+        }
+
+        Ok(values)
+    }
+}
+
+#[async_trait::async_trait]
+impl Handler for SumPackages {
+    type Batch = Vec<Self::Value>;
+
+    fn batch(batch: &mut Self::Batch, values: Vec<Self::Value>) {
+        batch.extend(values);
+    }
+
+    async fn commit(values: &Self::Batch, conn: &mut db::Connection<'_>) -> Result<usize> {
+        let updates = values.chunks(CHUNK_ROWS).map(|chunk| {
+            diesel::insert_into(sum_packages::table)
+                .values(chunk)
+                .on_conflict(sum_packages::package_id)
+                .do_update()
+                .set((
+                    sum_packages::package_version.eq(excluded(sum_packages::package_version)),
+                    sum_packages::move_package.eq(excluded(sum_packages::move_package)),
+                    sum_packages::cp_sequence_number.eq(excluded(sum_packages::cp_sequence_number)),
+                ))
+                .execute(conn)
```

Comment:
looks like we cant use `AsChangeset` when dealing with multiple writes huh

Reply:
Hmm, not sure -- maybe there's an `as_change_set()` or something? But the thing is that in this case we would only use the change set type in one place -- here -- so there's not a lot of value in introducing it.